service-monitor/src/main/java/com/yahoo/vespa/service/duper/DuperModel.java

```diff
@@ -20,37 +25,90 @@
 public class DuperModel {
     private static Logger logger = Logger.getLogger(DuperModel.class.getName());
 
-    private final Map<ApplicationId, ApplicationInfo> applications = new TreeMap<>();
+    private final Map<ApplicationId, ApplicationInfo> applicationsById = new HashMap<>();
+    private final Map<HostName, ApplicationInfo> applicationsByHostname = new HashMap<>();
+    private final Map<ApplicationId, Set<HostName>> hostnamesById = new HashMap<>();
+
     private final List<DuperModelListener> listeners = new ArrayList<>();
     private boolean isComplete = false;
 
     public void registerListener(DuperModelListener listener) {
-        applications.values().forEach(listener::applicationActivated);
+        applicationsById.values().forEach(listener::applicationActivated);
         listeners.add(listener);
     }
 
-    public void setCompleteness(boolean isComplete) { this.isComplete = isComplete; }
+    void setComplete() {
+        if (!isComplete) {
+            logger.log(LogLevel.INFO, "Bootstrap done - duper model is complete");
+            isComplete = true;
+
+            listeners.forEach(DuperModelListener::bootstrapComplete);
+        }
+    }
+
     public boolean isComplete() { return isComplete; }
 
+    public int numberOfApplications() {
+        return applicationsById.size();
+    }
+
+    public int numberOfHosts() {
+        return applicationsByHostname.size();
+    }
+
     public boolean contains(ApplicationId applicationId) {
-        return applications.containsKey(applicationId);
+        return applicationsById.containsKey(applicationId);
+    }
+
+    public Optional<ApplicationInfo> getApplicationInfo(ApplicationId applicationId) {
+        return Optional.ofNullable(applicationsById.get(applicationId));
+    }
+
+    public Optional<ApplicationInfo> getApplicationInfo(HostName hostName) {
+        return Optional.ofNullable(applicationsByHostname.get(hostName));
+    }
+
+    public List<ApplicationInfo> getApplicationInfos() {
+        return List.copyOf(applicationsById.values());
     }
 
     public void add(ApplicationInfo applicationInfo) {
-        applications.put(applicationInfo.getApplicationId(), applicationInfo);
-        logger.log(LogLevel.DEBUG, "Added " + applicationInfo.getApplicationId());
+        ApplicationInfo oldApplicationInfo = applicationsById.put(applicationInfo.getApplicationId(), applicationInfo);
+
+        final String logPrefix;
+        if (oldApplicationInfo == null) {
+            logPrefix = isComplete ? "New application " : "Bootstrapped application ";
+        } else {
+            logPrefix = isComplete ? "Reactivated application " : "Rebootstrapped application ";
+        }
+        logger.log(LogLevel.INFO, logPrefix + applicationInfo.getApplicationId());
+
+        Set<HostName> oldHostnames = hostnamesById.remove(applicationInfo.getApplicationId());
```

Comment:
Something like the following?

```
        HashSet<HostName> hostnames = hostnamesById
                .computeIfAbsent(applicationInfo.getApplicationId(), k -> new HashSet<>());

        Set<HostName> newHostnames;
        Set<HostName> removedHostnames;
        {
            newHostnames = applicationInfo.getModel().getHosts().stream()
                    .map(HostInfo::getHostname)
                    .map(HostName::from)
                    .collect(Collectors.toCollection(HashSet::new));

            removedHostnames = new HashSet<>(hostnames);
            removedHostnames.removeAll(newHostnames);

            newHostnames.removeAll(hostnames);
        }

        // These corrects applicationById
        hostnames.removeAll(removedHostnames);
        hostnames.addAll(newHostnames);

        // These corrects applicationByHostname
        removedHostnames.forEach(applicationsByHostname::remove);
        newHostnames.forEach(hostname -> applicationsByHostname.put(hostname, applicationInfo));
```

This will create more objects (1 additional HashSet) and will have one additional O(n) operation (removeAll) compared to current.

In addition, I would have to make applicationsByHostname map to ApplicationId instead of ApplicationInfo, otherwise unchanged hostnames may point to old ApplicationInfo. Which would incur a double map lookup to get ApplicationInfo instead of a single lookup.

Reply:
No, I was thinking something like this:
```
        Set<HostName> oldHostnames = hostnamesById.remove(applicationInfo.getApplicationId());
        Set<HostName> hostnames = applicationInfo.getModel().getHosts().stream()
                .map(HostInfo::getHostname)
                .map(HostName::from)
                .collect(Collectors.toSet());

        if (!Objects.equals(oldHostnames, hostnames)) {
            if (oldHostnames != null) {
                oldHostnames.forEach(applicationsByHostname::remove);
            }

            hostnamesById.put(applicationInfo.getApplicationId(), hostnames);
            hostnames.forEach(hostname -> applicationsByHostname.put(hostname, applicationInfo));
        }
```

Reply:
I now have a version I think is just as effective in the no-change case, but also performs well in the other cases. I have also added a log warning on duplicate host allocation.